From 3b4ee2509b0550d9cb01f05ac0b97e8f9c2107cc Mon Sep 17 00:00:00 2001
From: rajeshvenkata <rajesh.venkat.p@gmail.com>
Date: Fri, 30 Jan 2026 00:00:00 +0000
Subject: [PATCH 5/5] Fix ARM64 iPXE build and bump ipxe commit

This patch:
- Bumps iPXE commit to 9d4a2ee for latest drivers and fixes
- Fixes ARM64 build issues with older GCC versions by removing
  unsupported compiler flags (-Wno-dangling-pointer, -Wno-address-of-packed-member)
- Fixes ARM64 assembly constraint issues in bigint.h
- Disables patch_bnxt_rx_buffers which causes illegal byte sequence errors

Signed-off-by: rajeshvenkata <rajesh.venkat.p@gmail.com>
---
 .../ipxe/binary/file/script/build_ipxe.sh         | 15 ++++++++++++---
 smee/internal/ipxe/binary/file/script/ipxe.commit |  2 +-
 2 files changed, 13 insertions(+), 4 deletions(-)

diff --git a/smee/internal/ipxe/binary/file/script/build_ipxe.sh b/smee/internal/ipxe/binary/file/script/build_ipxe.sh
index cba99b23..d8073b92 100755
--- a/smee/internal/ipxe/binary/file/script/build_ipxe.sh
+++ b/smee/internal/ipxe/binary/file/script/build_ipxe.sh
@@ -88,8 +88,17 @@ function copy_custom_files() {
 # see http://lists.ipxe.org/pipermail/ipxe-devel/2018-August/006254.html .
 function customize_aarch_build() {
     local ipxe_dir="$1"
-    # http://lists.ipxe.org/pipermail/ipxe-devel/2018-August/006254.html
-    sed -i.bak '/^WORKAROUND_CFLAGS/ s|^|#|' "${ipxe_dir}"/src/arch/arm64/Makefile
+    # Fix compiler flag compatibility for older ARM64 GCC versions
+    sed -i 's/-Wno-dangling-pointer//g' "${ipxe_dir}"/src/Makefile.housekeeping
+    sed -i 's/-Wno-address-of-packed-member//g' "${ipxe_dir}"/src/Makefile.housekeeping
+    # Apply minimal ARM64 assembly constraint fixes directly to the bigint.h file
+    local bigint_h="${ipxe_dir}/src/arch/arm64/include/bits/bigint.h"
+    if [ -f "${bigint_h}" ]; then
+        sed -i 's/"=@cccs" ( carry )/"=r" ( carry )/g' "${bigint_h}"
+        sed -i 's/"=@cccc" ( borrow )/"=r" ( borrow )/g' "${bigint_h}"
+    fi
+    # Original ARM64 workaround - http://lists.ipxe.org/pipermail/ipxe-devel/2018-August/006254.html
+    sed -i '/^WORKAROUND_CFLAGS/ s|^|#|' "${ipxe_dir}"/src/arch/arm64/Makefile
 }
 
 # Workaround for Broadcom NetXtreme driver bug that causes a hang when
@@ -97,7 +106,7 @@ function customize_aarch_build() {
 # https://github.com/ipxe/ipxe/issues/1023#issuecomment-1898585257
 function patch_bnxt_rx_buffers() {
     local ipxe_dir="$1"
-    sed -i 's/\(#define NUM_RX_BUFFERS \).*/\12/' "${ipxe_dir}"/src/drivers/net/bnxt/bnxt.h
+    # Skip this patch as it's causing illegal byte sequence errors
 }
 
 # customize orchestrates the process for adding custom headers to an ipxe compile.
diff --git a/smee/internal/ipxe/binary/file/script/ipxe.commit b/smee/internal/ipxe/binary/file/script/ipxe.commit
index 94c98200..b9348c9b 100644
--- a/smee/internal/ipxe/binary/file/script/ipxe.commit
+++ b/smee/internal/ipxe/binary/file/script/ipxe.commit
@@ -1 +1 @@
-8460dc4e8ffc98db62377d1c5502d6aac40f5a64
+9d4a2ee3538f28a21a77c55c272c84b4e346dd92
-- 
2.49.0

